geo/src/algorithm/line_measures/metric_spaces/haversine.rs

```diff
@@ -174,13 +199,114 @@ impl<F: CoordFloat + FromPrimitive> InterpolatePoint<F> for Haversine {
         max_distance: F,
         include_ends: bool,
     ) -> impl Iterator<Item = Point<F>> {
-        crate::algorithm::HaversineIntermediate::haversine_intermediate_fill(
-            &start,
-            &end,
-            max_distance,
-            include_ends,
-        )
-        .into_iter()
+        let calculation = HaversineIntermediateFillCalculation::new(start, end);
+        let HaversineIntermediateFillCalculation { d, .. } = calculation;
+
+        let total_distance = d * F::from(MEAN_EARTH_RADIUS).unwrap();
+
+        if total_distance <= max_distance {
+            return if include_ends {
+                vec![start, end].into_iter()
+            } else {
+                vec![].into_iter()
+            };
+        }
+
+        let number_of_points = (total_distance / max_distance).ceil();
+        let interval = F::one() / number_of_points;
+
+        let mut current_step = interval;
+        let mut points = if include_ends { vec![start] } else { vec![] };
+
+        while current_step < F::one() {
+            let point = calculation.point_at_ratio(current_step);
+            points.push(point);
+            current_step = current_step + interval;
+        }
+
+        if include_ends {
+            points.push(end);
+        }
+
+        points.into_iter()
+    }
+}
+
+#[allow(clippy::many_single_char_names)]
+struct HaversineIntermediateFillCalculation<T> {
```

Comment:
The logic from HaversineIntermediateFillCalculation  was cut/paste from

https://github.com/georust/geo/pull/1222/files#diff-634e7da640fe8307cf8b1ef08ca56b7bfd954216ed7a41bee8e5d06b2e660d1fL94

Though I did re-organize the methods, which were previously free methods, now instance methods. The actual implementation logic is identical.